src/Compilers/CSharp/Portable/Binder/Binder.cs

```diff
@@ -705,6 +705,24 @@ internal static ObsoleteDiagnosticKind ReportDiagnosticsIfObsoleteInternal(Diagn
             return kind;
         }
 
+        internal static void ReportDiagnosticsIfUnmanagedCallersOnly(DiagnosticBag diagnostics, MethodSymbol symbol, Location location, bool isDelegateConversion)
+        {
+            var unmanagedCallersOnlyAttributeData = symbol.UnmanagedCallersOnlyAttributeData;
+            if (unmanagedCallersOnlyAttributeData != null)
+            {
+                // Either we haven't yet bound the attributes of this method, or there is an UnmanagedCallersOnly present.
+                // In the former case, we use a lazy diagnostic that may end up being ignored later, to avoid causing a
+                // binding cycle.
+                diagnostics.Add(unmanagedCallersOnlyAttributeData == UnmanagedCallersOnlyAttributeData.Uninitialized
+                                    ? (DiagnosticInfo)new LazyUnmanagedCallersOnlyMethodCalledDiagnosticInfo(symbol, isDelegateConversion)
```

Comment:
Do we need a lazy diagnostic? Could we just check if the attribute is present and create the `CSDiagnosticInfo` immediately?

Reply:
Obsolete is bound early, but still uses a lazy diagnostic. I'm assuming that if it was able to get away without a lazy diagnostic, it would have.

---
In reply to: [483713066](https://github.com/dotnet/roslyn/pull/47311#discussion_r483713066) [](ancestors = 483713066)